examples/external_decorators/run.py

```diff
@@ -0,0 +1,241 @@
+        decorators: dict[str, list[ast.Call]],
+        imports,
+        imports_from,
+    ) -> Any:
+        """Decorate an ast using a decorator mapping {function name: decorators}"""
+        self.module_docstring = module_docstring
+        self.imports = imports
+        self.imports_from = imports_from
+        self.decorators = decorators
+        return self.visit(tree)
+
+    def visit_Module(self, node):
+        for import_from_ in self.imports_from[::-1]:
+            node.body.insert(0, import_from_)
+
+        for import_ in self.imports[::-1]:
+            node.body.insert(0, import_)
+
+        if self.module_docstring:
+            node.body.insert(0, ast.Expr(ast.Constant(self.module_docstring)))
+
+        return node
+
+    def visit_FunctionDef(self, node):
+        """copies a function node and decorates it using `decorators` passed to .decorate()
+        Note. The added decorators will be "on top" of existing decorators, in the order passed.
+        """
+        new_decorator_list = self.decorators.get(node.name, []) + node.decorator_list
+
+        new_node = ast.copy_location(
+            ast.FunctionDef(
+                name=node.name,
+                args=node.args,
+                body=node.body,
+                decorator_list=new_decorator_list,
+                returns=node.returns,
+            ),
+            node,
+        )
+        return new_node
+
+
+# Read the script main to understand this function step by step
+def use_external_decorators(
+    undecorated_module: Union[str, ModuleType], external_decorators_module: Union[str, ModuleType]
+) -> ModuleType:
+    def _get_module_path(module):
+        if isinstance(module, ModuleType):
+            if not module.__spec__.has_location:
+                raise FileNotFoundError
+
+            module_path = module.__spec__.origin
+
+        elif isinstance(module, str):
+            module_path = module
+
+        else:
+            raise FileNotFoundError
+
+        return module_path
+
+    undecorated_module_path = _get_module_path(undecorated_module)
+    external_decorators_module_path = _get_module_path(external_decorators_module)
+
+    with open(undecorated_module_path, "r") as f:
+        undecorated_ast = ast.parse(f.read())
+
+    with open(external_decorators_module_path, "r") as f:
+        external_decorators_ast = ast.parse(f.read())
+
+    reader = HamiltonDecoratorReader()
+    reader.visit(external_decorators_ast)
+
+    writer = HamiltonDecoratorWriter()
+    decorated_ast = writer.decorate(
+        undecorated_ast,
+        reader.module_docstring,
+        reader.decorators,
+        reader.imports,
+        reader.imports_from,
+    )
+
+    module_uuid = str(uuid.uuid4())
+    module_object = ModuleType(module_uuid, reader.module_docstring)
+    sys.modules[module_uuid] = module_object
+    exec(ast.unparse(decorated_ast), module_object.__dict__)
+
+    return module_object
+
+
+if __name__ == "__main__":
```

Comment:
Worth using `click` to break this into multiple (2) comands that each to a use-case. Then they can take it in as sys.argv.